vowpalwabbit/parser.cc

```diff
@@ -673,18 +662,11 @@ namespace VW
 example& get_unused_example(vw* all)
 {
   parser* p = all->p;
-  while (true)
-  {
-    std::unique_lock<std::mutex> lock(p->examples_lock);
-    if (p->examples[p->begin_parsed_examples % p->ring_size].in_use == false)
-    {
-      example& ret = p->examples[p->begin_parsed_examples++ % p->ring_size];
-      ret.in_use = true;
-      return ret;
-    }
-    else
-      p->example_unused.wait(lock);
-  }
+  std::unique_lock<std::mutex> lock(p->pool_lock);
+  auto ex = p->example_pool.get_object();
+  ex->in_use = true;
+  p->begin_parsed_examples++;
+  return *ex;
```

Comment:
I'm slightly worried about the semantics here.  What happens when the parser is faster than the learner as far as memory usage?

Reply:
If the parser is faster then the queue will fill until the limit given (ring_size) and wait to add any more until some has been drained

Reply:
Cool.